File: scripts/tool-benchmark/hard-suite.ts
import { killAllProcesses } from "../../packages/cli/src/tools/processes.js";
import {
  ALTERNATION_PATTERN,
  FANOUT,
  hardIterations,
  MISS_PATTERN,
  samplesFor,
  UNICODE_CHUNKS,
  UNICODE_UNIT,
  UNICODE_UNIT_REPEATS,
  type Weight,
} from "./config.js";
import { hardCases, packagePath, resetCase, unitPath } from "./hard-cases.js";
import { type CaseResult, call, elapsed, getProcessId, measurer } from "./harness.js";

/**
 * Times the hard cases, then the two shapes that are not a single call.
 *
 * The single-call cases come from `hard-cases.ts` so the verifier exercises the
 * exact arguments the timings came from. Concurrency and the unicode ring
 * buffer need a live process or a batch, and are built here.
 */

const INTERACTIVE_COMMAND = process.platform === "win32" ? "more" : "cat";

interface Context {
  root: string;
}

export async function runHardSuite(root: string, quick: boolean): Promise<CaseResult[]> {
  const context = { root };
  const measure = measurer(samplesFor("hard", quick));
  const count = (weight: Weight) => hardIterations(weight, quick);
  const cases: CaseResult[] = [];

  for (const testCase of hardCases()) {
    cases.push(
      await measure(testCase.name, count(testCase.weight), async () => {
        if (testCase.reset) await resetCase(root);
        return elapsed(() => call(context, testCase.tool, testCase.arguments));
      }),
    );
  }

  cases.push(await unicodeOutputCase(context, measure, count("light")));
  cases.push(...(await concurrencyCases(context, measure, count)));

  killAllProcesses();
  return cases;
}

/**
 * A full ring buffer of multi-byte text, read from the start.
 *
 * The cursor the contract hands out counts UTF-16 units, which is what a
 * JavaScript string is indexed in and what a Rust `String` is not.
 */
async function unicodeOutputCase(
  context: Context,
  measure: ReturnType<typeof measurer>,
  iterations: number,
): Promise<CaseResult> {
  const processId = await fillUnicodeRing(context);
  const result = await measure("get_command_output_unicode", iterations, () =>
    elapsed(() => call(context, "get_command_output", { processId, cursor: 0 })),
  );

  await call(context, "kill_command", { processId });
  killAllProcesses();
  return result;
}

/** Starts a process and overruns its 256,000-unit ring with multi-byte text. */
export async function fillUnicodeRing(context: Context): Promise<string> {
  const started = await call(context, "start_command", { command: INTERACTIVE_COMMAND });
  const processId = getProcessId(started);
  const chunk = UNICODE_UNIT.repeat(UNICODE_UNIT_REPEATS);

  for (let index = 0; index < UNICODE_CHUNKS; index++) {
    await call(context, "send_command_input", { processId, data: chunk, newline: false });
  }
  await new Promise((resolve) => setTimeout(resolve, 250));
  return processId;
}

/**
 * Bursts, because the relay does not wait for one call before starting the next.
 *
 * Timed as one batch: the question is how long the whole burst takes, not how
 * long any call inside it took.
 */
async function concurrencyCases(
  context: Context,
  measure: ReturnType<typeof measurer>,
  count: (weight: Weight) => number,
): Promise<CaseResult[]> {
  const reads = Array.from({ length: FANOUT.reads }, (_, index) => unitPath(index));
  const cases: CaseResult[] = [];

  cases.push(
    await measure(`concurrent_reads_x${FANOUT.reads}`, count("heavy"), () =>
      elapsed(() => Promise.all(reads.map((path) => call(context, "read_file", { path })))),
    ),
  );
  cases.push(
    await measure(`concurrent_greps_x${FANOUT.greps}`, count("xheavy"), () =>
      elapsed(() =>
        Promise.all(
          Array.from({ length: FANOUT.greps }, () =>
            call(context, "grep", {
              pattern: ALTERNATION_PATTERN,
              path: "corpus/pkg-00",
              caseInsensitive: true,
              maxResults: 200,
            }),
          ),
        ),
      ),
    ),
  );

  const mixed = FANOUT.mixedReads + FANOUT.mixedLists + FANOUT.mixedGreps;
  cases.push(
    await measure(`concurrent_mixed_x${mixed}`, count("xheavy"), () =>
      elapsed(() =>
        Promise.all([
          ...Array.from({ length: FANOUT.mixedReads }, (_, index) =>
            call(context, "read_file", { path: unitPath(index) }),
          ),
          ...Array.from({ length: FANOUT.mixedLists }, (_, index) =>
            call(context, "list_files", { path: packagePath(index), recursive: true }),
          ),
          ...Array.from({ length: FANOUT.mixedGreps }, (_, index) =>
            call(context, "grep", {
              pattern: MISS_PATTERN,
              path: packagePath(index),
              maxResults: 200,
            }),
          ),
        ]),
      ),
    ),
  );

  return cases;
}

File: crates/exeora-cli/examples/compare_tools/hard_suite.rs
//! The same ten tools, at sizes and shapes that stop being a syscall wrapper.

use crate::{
    hard_cases::{
        ALTERNATION_PATTERN, MISS_PATTERN, hard_cases, package_path, reset_case, unit_path,
    },
    harness::{CaseResult, Config, Weight, execute, measure, process_id, serialize, timed},
};
use exeora_cli::{protocol::ToolName, tools::ToolEngine};
use futures_util::future::join_all;
use serde_json::{Value, json};
use std::{
    path::Path,
    thread,
    time::{Duration, Instant},
};
use tokio::runtime::Runtime;
use tokio_util::sync::CancellationToken;

pub const UNICODE_UNIT: &str = "héllo→世界 ";
pub const UNICODE_UNIT_REPEATS: usize = 7_111;
pub const UNICODE_CHUNKS: usize = 8;

const FANOUT_READS: usize = 64;
const FANOUT_GREPS: usize = 8;
const MIXED_READS: usize = 32;
const MIXED_LISTS: usize = 8;
const MIXED_GREPS: usize = 8;

#[cfg(unix)]
pub const INTERACTIVE_COMMAND: &str = "cat";
#[cfg(windows)]
pub const INTERACTIVE_COMMAND: &str = "more";

pub fn run(runtime: &Runtime, engine: &ToolEngine, root: &Path, config: Config) -> Vec<CaseResult> {
    let mut cases = Vec::new();

    for hard in hard_cases() {
        cases.push(measure(
            hard.name,
            config.hard(hard.weight),
            config.samples,
            || {
                if hard.reset {
                    reset_case(root);
                }
                // Cloned outside the timed region: the TypeScript side hands the
                // same object to every call and is charged for no copy at all.
                let payload = hard.arguments.clone();
                timed(|| {
                    execute(runtime, engine, root, hard.tool, payload);
                })
            },
        ));
    }

    cases.push(unicode_output(runtime, engine, root, config));
    cases.extend(concurrency(runtime, engine, root, config));
    runtime.block_on(engine.kill_all());
    cases
}

/// A full ring buffer of multi-byte text, read from the start.
///
/// The cursor the contract hands out counts UTF-16 units, which is what a
/// JavaScript string is indexed in and what a Rust `String` is not.
fn unicode_output(
    runtime: &Runtime,
    engine: &ToolEngine,
    root: &Path,
    config: Config,
) -> CaseResult {
    let id = fill_unicode_ring(runtime, engine, root);
    let result = measure(
        "get_command_output_unicode",
        config.hard(Weight::Light),
        config.samples,
        || {
            timed(|| {
                execute(
                    runtime,
                    engine,
                    root,
                    ToolName::GetCommandOutput,
                    json!({"processId":id, "cursor":0}),
                );
            })
        },
    );

    execute(
        runtime,
        engine,
        root,
        ToolName::KillCommand,
        json!({"processId":id}),
    );
    runtime.block_on(engine.kill_all());
    result
}

/// Starts a process and overruns its 256,000-unit ring with multi-byte text.
pub fn fill_unicode_ring(runtime: &Runtime, engine: &ToolEngine, root: &Path) -> String {
    let started = execute(
        runtime,
        engine,
        root,
        ToolName::StartCommand,
        json!({"command":INTERACTIVE_COMMAND}),
    );
    let id = process_id(&started);
    let chunk = UNICODE_UNIT.repeat(UNICODE_UNIT_REPEATS);

    for _ in 0..UNICODE_CHUNKS {
        execute(
            runtime,
            engine,
            root,
            ToolName::SendCommandInput,
            json!({"processId":id, "data":chunk, "newline":false}),
        );
    }
    thread::sleep(Duration::from_millis(250));
    id
}

/// Bursts, because the relay does not wait for one call before starting the next.
fn concurrency(
    runtime: &Runtime,
    engine: &ToolEngine,
    root: &Path,
    config: Config,
) -> Vec<CaseResult> {
    let batch = |calls: &[(ToolName, Value)]| {
        let started = Instant::now();
        runtime.block_on(async {
            let futures = calls.iter().map(|(tool, arguments)| {
                engine.execute(root, *tool, arguments.clone(), CancellationToken::new())
            });
            for result in join_all(futures).await {
                serialize(&result.expect("concurrent call failed"));
            }
        });
        started.elapsed().as_nanos()
    };

    let reads: Vec<(ToolName, Value)> = (0..FANOUT_READS)
        .map(|index| (ToolName::ReadFile, json!({"path":unit_path(index)})))
        .collect();
    let greps: Vec<(ToolName, Value)> = (0..FANOUT_GREPS)
        .map(|_| {
            (
                ToolName::Grep,
                json!({"pattern":ALTERNATION_PATTERN, "path":"corpus/pkg-00", "caseInsensitive":true, "maxResults":200}),
            )
        })
        .collect();

    let mut mixed: Vec<(ToolName, Value)> = (0..MIXED_READS)
        .map(|index| (ToolName::ReadFile, json!({"path":unit_path(index)})))
        .collect();
    mixed.extend((0..MIXED_LISTS).map(|index| {
        (
            ToolName::ListFiles,
            json!({"path":package_path(index), "recursive":true}),
        )
    }));
    mixed.extend((0..MIXED_GREPS).map(|index| {
        (
            ToolName::Grep,
            json!({"pattern":MISS_PATTERN, "path":package_path(index), "maxResults":200}),
        )
    }));

    vec![
        measure(
            format!("concurrent_reads_x{FANOUT_READS}"),
            config.hard(Weight::Heavy),
            config.samples,
            || batch(&reads),
        ),
        measure(
            format!("concurrent_greps_x{FANOUT_GREPS}"),
            config.hard(Weight::XHeavy),
            config.samples,
            || batch(&greps),
        ),
        measure(
            format!("concurrent_mixed_x{}", mixed.len()),
            config.hard(Weight::XHeavy),
            config.samples,
            || batch(&mixed),
        ),
    ]
}
